src/server/generateRandomToken.js

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2021-2022 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/fluid-project/preferencesServer/blob/main/LICENSE
+ */
+
+"use strict";
+
+const crypto = require("crypto");
+
+/**
+ * Create a unpredicatable random string of the given length.  This can be used
+ * to create nonce and other such tokens.  The tokens generated are URL safe
+ * based on RFC 4648 (https://www.rfc-editor.org/rfc/rfc4648.html#section-5).
+ *
+ * @param {Integer} length - The length of the state string.
+ * @return {String} a random set of characters.
+ */
+function generateRandomToken(length) {
```

Comment:
Please cite the stackoverflow answer this code was cribbed from: https://stackoverflow.com/a/36008760

Reply:
I read a number of sources about creating a nonce and making it URL safe.  The stackoverflow article you cite wasn't one of them.  The "Random values in Base64 format" section of "[Random values in JavaScript and Node.js | Tom's Blog](https://blog.abelotech.com/posts/generate-random-values-nodejs-javascript/#random-values-in-base64-format)" was the basis for the code.  Also, RFC 4648, as cited in the comment, provided the justification for replacing '+' with '-' and '/' with '_'.

At an earlier point in development, the random token generated would occasionally include a string of trailing equal signs and Google's OAuth2 server would complain.  This link describes the Google Authenticator error condition and recommends removing trailing equal signs.  The term used there is "padding", see:
https://github.com/google/google-authenticator/wiki/Key-Uri-Format#secret

Aside: I just noticed the last `replace()` removes characters, if trailing equal signs are found.  Since it removes characters, the length of the string returned could be less than what was requested, although I think it's unlikely -- conversion to base64 tends to make strings longer.  Nonetheless, a way to fix that is to replace trailing equal signs with some other safe character, say underscore, instead of removing them.

Reply:
Thanks for these careful thoughts and detailed explanation. I will include as a code comment. Also fix the handling of trailing equal signs.

Reply:
@klown, I was reading some threads regarding how to handle trailing equal signs in tokens, [this issue report](https://github.com/keepassxreboot/keepassxc/issues/3255) brings up a couple of good points:

1. Google suggested to remove them;
2. To decode you have to add the padding in anyway.

This makes me think the better approach might be what's currently in the code to remove trailing equal signs. In the case the padding needs to be added back to the full length, it's possible to know how many signs need to add back. If trailing equal signs are replaced to another character, it would be difficult to know if these characters were originally generated or replaced from equal signs.

Reply:
@cindyli, I read the same [issue](https://github.com/keepassxreboot/keepassxc/issues/3255) as you did, but months ago.  That was the issue that led me to [google's wiki article](https://github.com/google/google-authenticator/wiki/Key-Uri-Format#secret), and explained why some instances of the `state` query parameter were not working when sending OAuth2 requests to google.  That's why I put in that line to remove of any trailing `=` (padding).

I'm fine with leaving the code the way it is (that's the way wrote it :-) ).  But, I'm curious about your point about decoding.  Yes, the padding has to be there for decoding, but why decode it?  Or rather, who needs to decode it?  Maybe I should re-read the issue and wiki article (?).  I thought the goal was to have a string of characters that were difficult to guess or predict and were url-safe.  It's not as if you are, for example, encoding an image that the receiver is going to decode to get the image back.

Reply:
> but why decode it? Or rather, who needs to decode it?

Sorry my reply wasn't clear enough. Right, in our use case, the decoding is not needed. Rather the concern of the decoding leads to the use of removing instead of replacing padding as a more generic solution, which is also why I lean towards this solution.

Reply:
No worries.  And +1 for general solutions.  Maybe down the road there will be a need for decoding and it will "just work".